internal/character/clara/talent.go

```diff
@@ -100,54 +101,53 @@ func (c *char) doCounter(attackerID key.TargetID) {
 				Targets:      []key.TargetID{attackerID},
 				DamageType:   model.DamageType_PHYSICAL,
 				AttackType:   model.AttackType_INSERT,
-				BaseDamage:   mainTargetDamage,
+				BaseDamage:   info.DamageMap{model.DamageFormula_BY_ATK: percent},
 				StanceDamage: 30.0,
 				EnergyGain:   5.0,
 			})
 
 			// enhanced counter, damage to adjacent
 			// NOTE: in-game wording = damage value is based on main target's def
 			// mhy memes ?
-			if c.engine.HasModifier(c.id, UltCounter) {
-				splashDamage := info.DamageMap{model.DamageFormula_BY_ATK: (talent[c.info.TalentLevelIndex()] + ultDmgBoost[c.info.UltLevelIndex()]) * 0.5}
-
+			if hasUlt {
 				c.engine.Attack(info.Attack{
 					Source:       c.id,
 					Targets:      c.engine.AdjacentTo(attackerID),
 					DamageType:   model.DamageType_PHYSICAL,
 					AttackType:   model.AttackType_INSERT,
-					BaseDamage:   splashDamage,
+					BaseDamage:   info.DamageMap{model.DamageFormula_BY_ATK: percent * 0.5},
 					StanceDamage: 30.0,
 				})
+
+				// counter done, decrease counter stack
+				c.engine.ExtendModifierCount(c.id, UltCounter, -1.0)
 			}
 		},
 		Source:     c.id,
 		Priority:   info.CharInsertAttackSelf,
 		AbortFlags: []model.BehaviorFlag{model.BehaviorFlag_STAT_CTRL, model.BehaviorFlag_DISABLE_ACTION},
 	})
-
-	// add marker modifier on enemy attacker
-	c.engine.AddModifier(attackerID, info.Modifier{
-		Name:   TalentMark,
-		Source: c.id,
-		State:  State{skillLevelIndex: c.info.SkillLevelIndex(), ultLevelIndex: c.info.UltLevelIndex()},
-	})
 }
 
-// helper fn to see if a counter is eligible
-//
-// - if clara is not targeted AND < E6 -> no counter
-//
-// - clara targeted OR E6 + winning E6 50/50 -> counter
 func (c *char) canCounter(e event.AttackEnd) bool {
-	isClara := false
+	// clara has ult counter stacks left
+	hasUlt := c.engine.HasModifier(c.id, UltCounter) && c.engine.ModifierStackCount(c.id, c.id, UltCounter) > 0
```

Comment:
only need one of these checks. Modifiers will be automatically removed when their count reaches 0. If you call `ModifierStackCount`, will return `0` iff you do not have the modifier. Current state is looping over Clara modifiers twice